### PR TITLE
[shardformer] merge shardformer to main

### PR DESCRIPTION
**Pull request by FrankLeeeee**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

N/A

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

This PR releases the shardformer feature to the main branch.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/Chat/coati/trainer/.sft.py.swp
Comment (ver217) [inline applications/Chat/coati/trainer/.sft.py.swp:1]: This should be removed

File: colossalai/shardformer/examples/shardformer_benchmark.py
Comment (ver217) [inline colossalai/shardformer/examples/shardformer_benchmark.py:1]: Should this be out of colossalai's src?

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/examples/shardformer_benchmark.py
Comment (FrankLeeeee) [reply colossalai/shardformer/examples/shardformer_benchmark.py:1]: > Should this be out of colossalai's src?

Plan to move it to examples when the feature is ready for public release, i.e. stable to use

**Review by ver217 — APPROVED** [review]

File: applications/Chat/coati/trainer/.sft.py.swp
Comment (FrankLeeeee) [reply applications/Chat/coati/trainer/.sft.py.swp:1]: Removed

**Pull request merged**